src/main/java/graphql/incremental/IncrementalExecutionResult.java

```diff
@@ -0,0 +1,111 @@
+package graphql.incremental;
+
+import graphql.ExecutionResult;
+import graphql.ExperimentalApi;
+import org.reactivestreams.Publisher;
+
+import javax.annotation.Nullable;
+import java.util.List;
+
+/**
+ * A result that is part of an execution that includes incrementally delivered data (data has been deferred of streamed).
+ * <p>
+ * For example, this query
+ * <pre>
+ * query {
+ *   person(id: "cGVvcGxlOjE=") {
+ *     ...HomeWorldFragment @defer(label: "homeWorldDefer")
+ *     name
+ *     films @stream(initialCount: 1, label: "filmsStream") {
+ *       title
+ *     }
+ *   }
+ * }
+ * fragment HomeWorldFragment on Person {
+ *   homeWorld {
+ *     name
+ *   }
+ * }
+ * </pre>
+ * Could result on an incremental response with the following payloads (in JSON format here for simplicity).
+ * <p>
+ * <b>Response 1, the initial response does not contain any deferred or streamed results.</b>
+ * <pre>
+ * {
+ *   "data": {
+ *     "person": {
+ *       "name": "Luke Skywalker",
+ *       "films": [{ "title": "A New Hope" }]
+ *     }
+ *   },
+ *   "hasNext": true
+ * }
+ * </pre>
+ *
+ * <b>Response 2, contains the defer payload and the first stream payload.</b>
+ * <pre>
+ * {
+ *   "incremental": [
+ *     {
+ *       "label": "homeWorldDefer",
+ *       "path": ["person"],
+ *       "data": { "homeWorld": { "name": "Tatooine" } }
+ *     },
+ *     {
+ *       "label": "filmsStream",
+ *       "path": ["person", "films", 1],
+ *       "items": [{ "title": "The Empire Strikes Back" }]
+ *     }
+ *   ],
+ *   "hasNext": true
+ * }
+ * </pre>
+ *
+ * <b>Response 3, contains the final stream payload. Note how "hasNext" is "false", indicating this is the final response.</b>
+ * <pre>
+ * {
+ *   "incremental": [
+ *     {
+ *       "label": "filmsStream",
+ *       "path": ["person", "films", 2],
+ *       "items": [{ "title": "Return of the Jedi" }]
+ *     }
+ *   ],
+ *   "hasNext": false
+ * }
+ * </pre>
+ *
+ * <p>
+ * This implementation is based on the state of <a href="https://github.com/graphql/graphql-spec/pull/742">Defer/Stream PR</a>
+ * More specifically at the state of this
+ * <a href="https://github.com/graphql/graphql-spec/commit/c630301560d9819d33255d3ba00f548e8abbcdc6">commit</a>
+ * <p>
+ * The execution behaviour should match what we get from running Apollo Server 4.9.5 with graphql-js v17.0.0-alpha.2
+ */
+@ExperimentalApi
+public interface IncrementalExecutionResult extends ExecutionResult {
```

Comment:
I really don't want to deal with this hierarchy during the execution engine, because hierarchies are completed. we should make sure that this only used when the overall result is populated.